# Werkcollege-opdrachten Week 1.3

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [1]:
import warnings
warnings.simplefilter('ignore')
import pandas as pd
import sqlite3

Zet het bestand go_sales_train.sqlite in een makkelijk te vinden map

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales_train.sqlite te maken.

In [2]:
sales_conn = sqlite3.connect("go_sales_train.sqlite")

Lees van de ingelezen go_sales_train-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- returned_item
- return_reason

In [7]:
sql_query = "SELECT * FROM sqlite_master WHERE type='table';"
table_sales = pd.read_sql_query(sql_query, sales_conn)
table_sales

type           name       tbl_name  rootpage  \
0   table        country        country       480   
1   table  order_details  order_details        11   
2   table   order_header   order_header        24   
3   table   order_method   order_method        27   
4   table        product        product        28   
5   table   product_line   product_line        17   
6   table   product_type   product_type        10   
7   table  retailer_site  retailer_site        18   
8   table  return_reason  return_reason        30   
9   table  returned_item  returned_item        31   
10  table   sales_branch   sales_branch         2   
11  table    sales_staff    sales_staff         3   

                                                  sql  
0   CREATE TABLE "country" (\n\t"COUNTRY_CODE"\tIN...  
1   CREATE TABLE "order_details" (\n\t"ORDER_DETAI...  
2   CREATE TABLE "order_header" (\n\t"ORDER_NUMBER...  
3   CREATE TABLE "order_method" (\n\t"ORDER_METHOD...  
4   CREATE TABLE "product" (\n\t"PRODUCT_NUMBER"\t...  
5   CREATE TABLE "product_line" (\n\t"PRODUCT_LINE...  
6   CREATE TABLE "product_type" (\n\t"PRODUCT_TYPE...  
7   CREATE TABLE "retailer_site" (\n\t"RETAILER_SI...  
8   CREATE TABLE "return_reason" (\n\t"RETURN_REAS...  
9   CREATE TABLE "returned_item" (\n\t"RETURN_CODE...  
10  CREATE TABLE "sales_branch" (\n\t"SALES_BRANCH...  
11  CREATE TABLE "sales_staff" (\n\t"SALES_STAFF_C...

Krijg je een "no such table" error? Dan heb je misschien met .connect() per ongeluk een leeg  databasebestand (.sqlite) aangemaakt. <u>Let op:</u> lees eventueel de informatie uit het Notebook van werkcollege 1.1b nog eens goed door.

Als je tijdens onderstaande opdrachten uit het oog verliest welke tabellen er allemaal zijn, kan je deze Pythoncode uitvoeren:

In [ ]:
sql_query = "SELECT name FROM sqlite_master WHERE type='table';"
#Vul dit codeblok verder in
pd.read_sql(sql_query, sales_conn)
#Op de puntjes hoort de connectie naar go_sales_train óf go_staff_train óf go_crm_train te staan.

name
0         country
1   order_details
2    order_header
3    order_method
4         product
5    product_line
6    product_type
7   retailer_site
8   return_reason
9   returned_item
10   sales_branch
11    sales_staff

erachter 

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [67]:
overzicht_product = table_sales[(table_sales['PRODUCTION_COST'] < 100) & (table_sales['PRODUCTION_COST'] > 50)]
overzicht_product[['PRODUCT_NAME', 'PRODUCTION_COST']]


PRODUCT_NAME  PRODUCTION_COST
0         TrailChef Deluxe Cook Set            85.11
1            TrailChef Double Flame            75.00
2                   Hibernator Lite            60.00
3                        Hibernator            86.00
4   Hibernator Self - Inflating Mat            78.55
5               Hibernator Camp Cot            65.33
6      Canyon Mule Climber Backpack            62.50
7           Granite Climbing Helmet            52.86
8             Husky Harness Extreme            55.00
9                    Granite Hammer            57.13
10                        Polar Ice            73.33
11                     Polar Sports            80.00
12                       Polar Wave            60.00
13                    Polar Extreme            72.50
14                     Edge Extreme            80.00
15                        Seeker 35            79.19
16                        Seeker 50            92.58
17                   Seeker Extreme            94.12
18                   Glacier Deluxe            56.25
19                      Glacier GPS            78.55
20                Blue Steel Putter            60.00
21            Blue Steel Max Putter            81.80
22              Course Pro Golf Bag            80.00

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [70]:
overzicht_product = table_sales[(table_sales['MARGIN'] < 0.20) | (table_sales['MARGIN'] > 0.60)]
overzicht_product[['PRODUCT_NAME', 'MARGIN']]



PRODUCT_NAME  MARGIN
0                Hibernator Pad    0.17
1  Canyon Mule Climber Backpack    0.17
2             BugShield Natural    0.67
3               BugShield Spray    0.67
4         BugShield Lotion Lite    0.70
5              BugShield Lotion    0.63
6             BugShield Extreme    0.63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [89]:


overzicht_product = table_sales[(table_sales['CURRENCY_NAME'] == 'francs')]
overzicht_product.sort_values(by='COUNTRY')

COUNTRY_CODE      COUNTRY LANGUAGE CURRENCY_NAME
0            17      Belgium       EN        francs
1             1       France       EN        francs
2             8  Switzerland       EN        francs

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [ ]:
overzicht_introductie_datum = table_sales[(table_sales['MARGIN'] > 0.50)]
overzicht_introductie_datum.sort_values(by='INTRODUCTION_DATE')

PRODUCT_NUMBER INTRODUCTION_DATE  PRODUCT_TYPE_CODE  PRODUCTION_COST  \
0              86        1995-02-15                 15             2.00   
1              67        2000-10-26                 10           116.19   
2             102        2003-12-10                 18           380.95   
3             111        2003-12-15                 20            81.80   
4             104        2003-12-18                 18           333.32   
5             105        2003-12-27                 19           555.90   
6             107        2004-01-13                 19           509.07   

   MARGIN    PRODUCT_IMAGE LANGUAGE                       PRODUCT_NAME  \
0    0.67   P86OP4IR15.jpg       EN                  BugShield Natural   
1    0.51   P67PA3WT10.jpg       EN               Mountain Man Extreme   
2    0.51  P102GE5IR18.jpg       EN           Hailstorm Titanium Irons   
3    0.55  P111GE5PT20.jpg       EN              Blue Steel Max Putter   
4    0.51  P102GE5IR18.jpg       EN      Lady Hailstorm Titanium Irons   
5    0.55  P105GE5WD19.jpg       EN       Hailstorm Titanium Woods Set   
6    0.55  P105GE5WD19.jpg       EN  Lady Hailstorm Titanium Woods Set   

                                         DESCRIPTION  
0  Contains no chemicals. Repels mosquitoes, blac...  
1  Solid titanium case, crown, bezel.  Combinatio...  
2  Made entirely of pure titanium. The ultimate i...  
3  Putter head is composed from a single piece of...  
4  Made entirely of pure titanium. A wider sole a...  
5  Crafted of 6-4 titanium which is 40% lighter a...  
6  Crafted of 6-4 titanium which is 40% lighter a...

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 7 rijen)

In [ ]:
sql_query = "SELECT * FROM sales_branch"
sales_branch = pd.read_sql_query(sql_query, sales_conn)

overzicht_verkoop = sales_branch[(sales_branch['ADDRESS2'].notna()) & (sales_branch['REGION'].notna())]
overzicht_verkoop[['ADDRESS1', 'CITY']]

ADDRESS1              CITY
2                Singelgravenplein 4         Amsterdam
13  Prol. Paseo de la Reforma No. 51  Distrito Federal
14               202-2-3 Hyakunincho             Tokyo
15                     543-225 Asahi        Osaka City
16                  2315 Queen's Ave         Melbourne
18             Avenida Paulista, 333         São Paulo
24                    3 Albany Court        Birmingham

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [ ]:
sql_query = "SELECT * FROM country "
table_sales = pd.read_sql_query(sql_query, sales_conn)

overzicht_geld = table_sales[(table_sales['CURRENCY_NAME'].str.contains('dollar'))]
overzicht_geld[['COUNTRY']]

COUNTRY
2   United States
3          Canada
11         Taiwan
14      Australia

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [137]:
sql_query = "SELECT * FROM retailer_site"
table_sales = pd.read_sql_query(sql_query, sales_conn)

overzicht_geld = table_sales[(table_sales['POSTAL_ZONE'].str.startswith('D')) & (table_sales['ADDRESS1'].notna()) & (table_sales['ADDRESS2'].notna())]

overzicht_geld = overzicht_geld[['ADDRESS1', 'ADDRESS2', 'CITY']]
overzicht_geld



ADDRESS1       ADDRESS2       CITY
31  Röntgenstraße 90  3. Tür rechts  Frankfurt
59   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [154]:
sql_query = "SELECT * FROM returned_item"
table_sales = pd.read_sql_query(sql_query, sales_conn)


table_sales['RETURN_QUANTITY'].sum()

np.int64(11908)

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [158]:
sql_query = "SELECT * FROM sales_branch"
table_sales = pd.read_sql_query(sql_query, sales_conn)

table_sales['REGION'].nunique()



15

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [ ]:
sql_query = "SELECT * FROM product"
table_sales = pd.read_sql_query(sql_query, sales_conn)
table_sales

laagste_marge = table_sales['MARGIN'].min()
hoogste_marge = table_sales['MARGIN'].max()
gemiddelde_marge = round(table_sales['MARGIN'].mean(), 2)

resultaat_marge = pd.DataFrame({'Laagste': [laagste_marge],'Hoogste': [hoogste_marge],'Gemiddelde': [gemiddelde_marge]})

resultaat_marge


Laagste  Hoogste  Gemiddelde
0     0.17      0.7         0.4

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [178]:
geenadres = table_sales['ADDRESS2'].isna().sum()
print(geenadres)

286


Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [181]:
sql_query = "SELECT * FROM order_details"
table_sales = pd.read_sql_query(sql_query, sales_conn)

gemprijs = (table_sales['UNIT_SALE_PRICE'] < table_sales['UNIT_PRICE'])
producten = table_sales.loc[gemprijs, ['UNIT_COST']].mean()
print(producten)

UNIT_COST    127.340879
dtype: float64


Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [186]:
sql_query = "SELECT * FROM sales_staff "
table_sales = pd.read_sql_query(sql_query, sales_conn)

table_sales.groupby('POSITION_EN', as_index=False)['SALES_STAFF_CODE'].count()



POSITION_EN  SALES_STAFF_CODE
0                Branch Manager                19
1              District Manager                 4
2               General Manager                 1
3  Level 1 Sales Representative                13
4  Level 2 Sales Representative                39
5  Level 3 Sales Representative                24
6              Regional Manager                 2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [187]:
werknemers = table_sales.groupby('WORK_PHONE', as_index=False)['SALES_STAFF_CODE'].count()
werknemers.loc[werknemers['SALES_STAFF_CODE'] > 4]

WORK_PHONE  SALES_STAFF_CODE
4     +(44) 121 3505267                 5
10  +(61) 03 2982  4242                 5
11  +31 (0)20 692 93 94                 5
13    +33 1 68 94 52 20                 5
18     1 (206) 292-0012                 5
19     1 (305) 557-4810                 6
20     1 (310) 281-5722                 5
21     1 (403) 232-5986                 5
22     1 (416) 493-5595                 5
23     1 (617) 268-6754                 6

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [193]:
sql_query = "SELECT * FROM country"
sql_query1 = "SELECT * FROM retailer_site"
table_sales = pd.read_sql_query(sql_query1, sales_conn)
country = pd.read_sql_query(sql_query, sales_conn)

merge = pd.merge(table_sales, country, left_on='COUNTRY_CODE', right_on='COUNTRY_CODE', how='inner')
nederlands = merge['COUNTRY'] == 'Netherlands'
merge.loc[nederlands, ['ADDRESS1', 'CITY']]

ADDRESS1          CITY
240                   Beets 36         Beets
241               Perestraat 1      Deventer
242  Laan van Meerdervoort 966      Den Haag
243             Kalfjeslaan 47    Amstelveen
244               Startbaan 11    Amstelveen
245      Noorderspoorsingel 28     Groningen
246              Hogehilweg 13     Amsterdam
247             Spoorstraat 43    Varsseveld
248             Westhavenweg 8     Amsterdam
249              Westzeedijk 4        Muiden
250            Gooiergracht 91         Laren
251            Hoofdstraat 150   Pieterburen
252                Olde Diek 2         Laren
253     Amsterdamsestraatweg 8      Maarssen
254        Krayenhoffstraat 11     Amsterdam
255         Overschiestraat 19     Amsterdam
256              Havenbuurt 12        Marken
257               Duivendaal 1    Wageningen
258              Voorstraat 37  Nieuweschans
259         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [194]:
sql_query = "SELECT * FROM product"
sql_query1 = "SELECT * FROM product_type"
product = pd.read_sql_query(sql_query, sales_conn)
product_type = pd.read_sql_query(sql_query1, sales_conn)

merge = pd.merge(product, product_type, left_on='PRODUCT_TYPE_CODE', right_on='PRODUCT_TYPE_CODE', how='inner')
eyewear = merge['PRODUCT_TYPE_EN'] == 'Eyewear'
merge.loc[eyewear, ['PRODUCT_NAME']]

PRODUCT_NAME
67      Polar Sun
68      Polar Ice
69   Polar Sports
70     Polar Wave
71  Polar Extreme

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [205]:
sql_query = "SELECT * FROM retailer_site"
sql_query1 = "SELECT * FROM order_header"
sql_query2 = "SELECT * FROM sales_staff"

retailer_site = pd.read_sql_query(sql_query, sales_conn)
order_header = pd.read_sql_query(sql_query1, sales_conn)
sales_staff = pd.read_sql_query(sql_query2, sales_conn)


merge = pd.merge(retailer_site, order_header, left_on='RETAILER_SITE_CODE', right_on='RETAILER_SITE_CODE', how='inner')
merge = pd.merge(merge, sales_staff, left_on='SALES_STAFF_CODE', right_on='SALES_STAFF_CODE', how='inner')
conditions = merge['POSITION_EN'] == 'Branch Manager'
merge.loc[conditions, ['ADDRESS1', 'FIRST_NAME', 'LAST_NAME']].drop_duplicates()


ADDRESS1 FIRST_NAME LAST_NAME
3658  Alameda Santos, 9876     Bayard     Lopes

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 7 rijen) 

In [196]:
merge = pd.merge(sales_staff, order_header, left_on='SALES_STAFF_CODE', right_on='SALES_STAFF_CODE', how='outer')
condition = merge['POSITION_EN'].str.contains('Manager')
merge.loc[condition, ['POSITION_EN', 'ORDER_DATE']].drop_duplicates()

POSITION_EN  ORDER_DATE
0       Branch Manager         NaN
320   Regional Manager         NaN
648   District Manager         NaN
3748    Branch Manager  2024-09-09
3749    Branch Manager  2024-03-13
3750    Branch Manager  2024-01-09
4186   General Manager         NaN

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [197]:
sql_query3 = "SELECT * FROM order_details" 
order_details = pd.read_sql_query(sql_query3, sales_conn)

merge = pd.merge(order_details, product, on='PRODUCT_NUMBER', how='inner')
merge = pd.merge(merge, product_type, on='PRODUCT_TYPE_CODE', how='inner')
condition = merge['QUANTITY'] > 750
merge.loc[condition, ['PRODUCT_NAME', 'PRODUCT_TYPE_EN']].drop_duplicates()

PRODUCT_NAME    PRODUCT_TYPE_EN
539        BugShield Extreme  Insect Repellents
1759                Star Peg              Tents
20121  BugShield Lotion Lite  Insect Repellents
20147         Sun Shelter 15          Sunscreen
20152         Sun Shelter 30          Sunscreen
31236        BugShield Spray  Insect Repellents
37272              Firefly 2           Lanterns
37348       BugShield Lotion  Insect Repellents
37363             Sun Shield          Sunscreen

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [198]:
product_order = pd.merge(product, order_details, on='PRODUCT_NUMBER', how='inner')
condition = ((product_order['UNIT_PRICE'] - product_order['UNIT_SALE_PRICE']) / product_order['UNIT_PRICE']) > 0.4
product_order.loc[condition, ['PRODUCT_NAME']].drop_duplicates()

PRODUCT_NAME
27830             BugShield Natural
28283               BugShield Spray
28694         BugShield Lotion Lite
29127              BugShield Lotion
29464             BugShield Extreme
30562                Sun Shelter 15
30929                Sun Shelter 30
34441  Hailstorm Titanium Woods Set

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [199]:
sql_query4 = "SELECT * FROM returned_item" 
returned_item = pd.read_sql_query(sql_query4, sales_conn)
sql_query5 = "SELECT * FROM return_reason" 
return_reason = pd.read_sql_query(sql_query5, sales_conn)

merge = pd.merge(returned_item, order_details, on='ORDER_DETAIL_CODE', how='inner')
merge = pd.merge(merge, return_reason, on='RETURN_REASON_CODE', how='inner')
condition = (merge['RETURN_QUANTITY'] / merge['QUANTITY']) > 0.9
merge.loc[condition, ['RETURN_DESCRIPTION_EN']].drop_duplicates()

RETURN_DESCRIPTION_EN
0  Unsatisfactory product
2   Wrong product shipped
3   Wrong product ordered

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [200]:
merge = pd.merge(product, product_type, on='PRODUCT_TYPE_CODE', how='inner')
merge.groupby('PRODUCT_TYPE_EN', as_index=False)['PRODUCT_NUMBER'].count()

PRODUCT_TYPE_EN  PRODUCT_NUMBER
0             Binoculars               4
1   Climbing Accessories               7
2           Cooking Gear              10
3                Eyewear               5
4              First Aid               5
5       Golf Accessories               4
6      Insect Repellents               5
7                  Irons               4
8                 Knives               5
9               Lanterns              12
10            Navigation               4
11                 Packs               6
12               Putters               3
13                  Rope               4
14                Safety               4
15         Sleeping Bags               7
16             Sunscreen               5
17                 Tents               6
18                 Tools               6
19               Watches               5
20                 Woods               4

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [201]:
merge = pd.merge(retailer_site, country, on='COUNTRY_CODE', how='inner')
merge.groupby('COUNTRY', as_index=False)['RETAILER_SITE_CODE'].count()

COUNTRY  RETAILER_SITE_CODE
0        Australia                  10
1          Austria                  10
2          Belgium                  10
3           Brazil                   1
4           Canada                  31
5            China                  10
6          Denmark                   6
7          Finland                   5
8           France                  29
9          Germany                  32
10           Italy                  12
11           Japan                  25
12           Korea                  10
13          Mexico                  10
14     Netherlands                  20
15           Spain                   9
16          Sweden                  12
17     Switzerland                  13
18          Taiwan                  10
19  United Kingdom                  25
20   United States                 101

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [202]:
merge = pd.merge(product_type, product, on='PRODUCT_TYPE_CODE', how='inner')
merge = pd.merge(merge, order_details, on='PRODUCT_NUMBER', how='inner')
condition = merge['PRODUCT_TYPE_EN'] == 'Cooking Gear'
merge = merge.loc[condition, :]
merge.groupby('PRODUCT_NAME', as_index=False).agg(Totaalhoeveelheid = ('QUANTITY', 'sum'), Gemverkoop = ('UNIT_SALE_PRICE', 'mean')).sort_values('Totaalhoeveelheid')


PRODUCT_NAME  Totaalhoeveelheid  Gemverkoop
3  TrailChef Deluxe Cook Set               3110  121.217638
7     TrailChef Single Flame               8504   63.827622
5           TrailChef Kettle               8826   12.201235
1         TrailChef Cook Set              10878   51.687926
4     TrailChef Double Flame              12476  132.672882
8         TrailChef Utensils              13776   17.294151
6      TrailChef Kitchen Kit              22206   22.345975
0          TrailChef Canteen              26544   11.979418
2              TrailChef Cup              27418    6.976490
9        TrailChef Water Bag              36738    6.032506

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [203]:
merge = pd.merge(sales_branch, country, on='COUNTRY_CODE', how='outer')
merge = pd.merge(merge, retailer_site, on='COUNTRY_CODE', how='outer')
merge.groupby('COUNTRY', as_index=False).agg(Klanten = ('RETAILER_CODE', 'count'))

COUNTRY  Klanten
0        Australia       10
1          Austria       10
2          Belgium       10
3           Brazil        1
4           Canada       62
5            China       10
6          Denmark        6
7          Finland        5
8           France       58
9          Germany       64
10           Italy       12
11           Japan       50
12           Korea       10
13          Mexico       10
14     Netherlands       20
15           Spain        9
16          Sweden       12
17     Switzerland       13
18          Taiwan       10
19  United Kingdom       50
20   United States      404

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [206]:
zonder_Verkoop = pd.merge(sales_staff, order_header, on='SALES_STAFF_CODE', how='left', indicator=True)
zonder_Verkoop = zonder_Verkoop[zonder_Verkoop['_merge'] == 'left_only']
zonder_Verkoop[['FIRST_NAME', 'LAST_NAME']]

FIRST_NAME      LAST_NAME
0         Denis           Pagé
320        Else         Mörike
321       Frank          Fuchs
647       Fritz         Hirsch
648        Jörg          Kunze
728       Maria      Iacobucci
946        Kick        Kalkman
1309      Karin      Bergström
1395      Sally          White
1421      Frank        Bretton
1422    Georges  Saint-Germain
1719       Alex      Rodriguez
2088   Samantha          Floyd
2392      Julie          Olsen
2530      Corey         Wright
2873    Estelle        Johnson
2874       Greg        Belding
2999      Julie        Béchard
3157      Lucía          Reyna
3401     Kazumi        Uragome
3559      Alice         Walter
4185     Walter         Taylor
4186   Penelope        Tamrine
4349      Polly         Linton
4590      Gracy        Gellens

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [212]:
condition = product['MARGIN'] < product['MARGIN'].mean()
products = product.loc[condition,['MARGIN']]
pd.DataFrame([products.count(), products.mean()])

MARGIN
0  31.000000
1   0.253226

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [ ]:
sql_query5 = "SELECT * FROM product" 
product1 = pd.read_sql_query(sql_query5, sales_conn) # ergens een query die runt en die het breekt "product 1"

merged = pd.merge(order_details, returned_item, on='ORDER_DETAIL_CODE', how='outer')
merged = pd.merge(merged, product1, on='PRODUCT_NUMBER', how='outer') # ergens een query die runt en die het breekt "product 1"
condition_price = merged['UNIT_SALE_PRICE'] > 500
condition_returned = merged['RETURN_DATE'].isna()
merged.loc[condition_price & condition_returned, ['PRODUCT_NAME']].drop_duplicates()



PRODUCT_NAME
3959                           Star Dome
4356                        Star Gazer 2
4699                        Star Gazer 3
5085                        Star Gazer 6
15041                     Husky Rope 200
33162              Hailstorm Steel Irons
33536           Hailstorm Titanium Irons
33900         Lady Hailstorm Steel Irons
34174      Lady Hailstorm Titanium Irons
34438       Hailstorm Titanium Woods Set
34733          Hailstorm Steel Woods Set
35118  Lady Hailstorm Titanium Woods Set
35397     Lady Hailstorm Steel Woods Set

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [217]:
for index, row in sales_staff.iterrows():
    position_en = sales_staff.at[index, 'POSITION_EN']
    if ('Manager' in position_en):
        sales_staff.at[index, 'IsManager'] = 'Ja'
    else:
        sales_staff.at[index, 'IsManager'] = 'Nee'
sales_staff

SALES_STAFF_CODE FIRST_NAME  LAST_NAME                   POSITION_EN  \
0                   4      Denis       Pagé                Branch Manager   
1                   5  Élizabeth     Michel  Level 3 Sales Representative   
2                   6      Émile   Clermont  Level 1 Sales Representative   
3                   7    Étienne     Jauvin  Level 2 Sales Representative   
4                  12    Elsbeth  Wiesinger  Level 2 Sales Representative   
..                ...        ...        ...                           ...   
97                120      Giele   Laermans  Level 1 Sales Representative   
98                121   François    De Crée  Level 1 Sales Representative   
99                122     Yvette    Lattrez  Level 3 Sales Representative   
100               123      Willi  Seefelder  Level 2 Sales Representative   
101               124     Sabine     Grüner  Level 3 Sales Representative   

            WORK_PHONE  EXTENSION                FAX                   EMAIL  \
0    +33 1 68 94 52 20      325.0  +33 1 68 94 56 60       DPage@grtd123.com   
1    +33 1 68 94 52 20      336.0  +33 1 68 94 56 60     EMichel@grtd123.com   
2    +33 1 68 94 52 20      378.0  +33 1 68 94 56 60   EClermont@grtd123.com   
3    +33 1 68 94 52 20      398.0  +33 1 68 94 56 60     EJauvin@grtd123.com   
4    +(49) 40 663 1990     1818.0  +(49) 40 663 4571  EWiesinger@grtd123.com   
..                 ...        ...                ...                     ...   
97     +32 16 20.73.21     1340.0    +32 16 20.73.32   GLaermans@grtd123.com   
98     +32 16 20.73.21     1642.0    +32 16 20.73.32     FDecree@grtd123.com   
99     +32 16 20.73.21     1633.0    +32 16 20.73.32    YLattrez@grtd123.com   
100  +(43) 13 79 56 32      325.0  +(43) 13 79 56 33  WSeefelder@grtd123.com   
101  +(43) 13 79 56 32      348.0  +(43) 13 79 56 33     SGruner@grtd123.com   

     DATE_HIRED  SALES_BRANCH_CODE IsManager  
0    1996-11-03                  6        Ja  
1    1995-06-08                  6       Nee  
2    1998-04-07                  6       Nee  
3    1997-08-16                  6       Nee  
4    1997-02-27                 13       Nee  
..          ...                ...       ...  
97   1999-11-15                 38       Nee  
98   1999-12-01                 38       Nee  
99   1999-12-09                 38       Nee  
100  1998-10-28                 39       Nee  
101  1998-02-18                 39       Nee  

[102 rows x 11 columns]

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [224]:
from datetime import date
year_now = date.today().year

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [223]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 

In [ ]:
from datetime import datetime
for index, row in sales_staff.iterrows():
    date_hired = sales_staff.at[index, 'DATE_HIRED']
    date_format = '%Y-%m-%d'
    date_obj = datetime.strptime(date_hired, date_format)

    verschil = year_now - date_obj.year

    if verschil < 12:
        sales_staff.at[index, 'Diensttijd'] = 'Kort in dienst'
    else:
        sales_staff.at[index, 'Diensttijd'] = 'Lang in dienst'
sales_staff.loc[:,['LAST_NAME', 'Diensttijd']]

LAST_NAME      Diensttijd
0         Pagé  Lang in dienst
1       Michel  Lang in dienst
2     Clermont  Lang in dienst
3       Jauvin  Lang in dienst
4    Wiesinger  Lang in dienst
..         ...             ...
97    Laermans  Lang in dienst
98     De Crée  Lang in dienst
99     Lattrez  Lang in dienst
100  Seefelder  Lang in dienst
101     Grüner  Lang in dienst

[102 rows x 2 columns]

## Van Jupyter Notebook naar Pythonproject

1. Richt de map waarin jullie tot nu toe hebben gewerkt in volgens de mappenstructuur uit de slides.
2. Maak van de ontstane mappenstructuur een Pythonproject dat uitvoerbaar is vanuit de terminal. Maak daarin een .py-bestand dat minstens 5 antwoorden uit dit notebook (in de vorm van een DataFrame) exporteert naar Excelbestanden. Alle notebooks mogen als notebook blijven bestaan.
3. Zorg ervoor dat dit Pythonproject zijn eigen repo heeft op Github. Let op: je virtual environment moet <b><u>niet</u></b> meegaan naar Github.

Je mag tijdens dit proces je uit stap 1 ontstane mappenstructuur aanpassen, zolang je bij het beoordelingsmoment kan verantwoorden wat de motivatie hierachter is. De slides verplichten je dus nergens toe.